File: src/main/java/org/jukeboxmc/inventory/FakeDoubleChestInventory.java
package org.jukeboxmc.inventory;

import com.nukkitx.nbt.NbtMap;
import com.nukkitx.protocol.bedrock.packet.BlockEntityDataPacket;
import com.nukkitx.protocol.bedrock.packet.ContainerOpenPacket;
import org.jetbrains.annotations.NotNull;
import org.jukeboxmc.Server;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;

import java.util.Arrays;
import java.util.List;

/**
 * @author LucGamesYT
 * @version 1.0
 */
public class FakeDoubleChestInventory extends FakeChestInventory {

    public FakeDoubleChestInventory() {
        super( null, -1, 27 * 2 );
    }

    public FakeDoubleChestInventory(String customName) {
        super( null, -1, 27 * 2 );
        this.customName = customName;
    }

    @Override
    public @NotNull InventoryType getType() {
        return InventoryType.FAKE_DOUBLE_CHEST;
    }

    @Override
    public void addViewer(@NotNull Player player, @NotNull Vector position, byte windowId ) {
        this.viewer.add( player );

        if ( this.openInventory.put( player, this ) != null ){
            return;
        }

        List<Vector> blockPosition = this.onOpenChest( player );
        this.blockPositions.put( player, blockPosition );

        Server.getInstance().getScheduler().scheduleDelayed( () -> {
            Vector vector = blockPosition.get( 0 );

            ContainerOpenPacket containerOpenPacket = new ContainerOpenPacket();
            containerOpenPacket.setId( windowId );
            containerOpenPacket.setUniqueEntityId( this.holderId );
            containerOpenPacket.setType( this.getWindowTypeId() );
            containerOpenPacket.setBlockPosition( vector.toVector3i() );

            player.getPlayerConnection().sendPacket( containerOpenPacket );

            this.sendContents( player );
        }, 3 );
    }

    @Override
    public @NotNull List<Vector> onOpenChest(@NotNull Player player ) {
        Vector positionA = new Vector( player.getBlockX(), player.getBlockY() + 2, player.getBlockZ() );
        Vector positionB = positionA.add( 1, 0, 0 );

        this.placeFakeChest( player, positionA );
        this.placeFakeChest( player, positionB );

        this.pair( player, positionA, positionB );
        this.pair( player, positionB, positionA );

        return Arrays.asList( positionA, positionB );
    }

    private void pair(@NotNull Player player, @NotNull Vector positionA, @NotNull Vector positionB ) {
        BlockEntityDataPacket blockEntityDataPacket = new BlockEntityDataPacket();
        blockEntityDataPacket.setBlockPosition( positionA.toVector3i() );
        blockEntityDataPacket.setData( this.toChestNBT( positionA, positionB ) );
        player.getPlayerConnection().sendPacket( blockEntityDataPacket );
    }

    private NbtMap toChestNBT(@NotNull Vector positionA, @NotNull Vector positionB ) {
        return NbtMap.builder()
                .putString( "id", "Chest" )
                .putInt( "x", positionA.getBlockX() )
                .putInt( "y", positionA.getBlockY() )
                .putInt( "z", positionA.getBlockZ() )
                .putInt( "pairx", positionB.getBlockX() )
                .putInt( "pairz", positionB.getBlockZ() )
                .putString( "CustomName", this.customName == null ? "Chest" : this.customName )
                .build();
    }

}

File: src/main/java/org/jukeboxmc/inventory/FakeDoubleChestInventory.kt
package org.jukeboxmc.inventory

import com.nukkitx.nbt.NbtMap
import com.nukkitx.protocol.bedrock.packet.BlockEntityDataPacket
import com.nukkitx.protocol.bedrock.packet.ContainerOpenPacket
import org.jukeboxmc.Server
import org.jukeboxmc.math.Vector
import org.jukeboxmc.player.Player

/**
 * @author LucGamesYT
 * @version 1.0
 */
class FakeDoubleChestInventory : FakeChestInventory {
    constructor() : super(null, -1, 27 * 2)
    constructor(customName: String?) : super(null, -1, 27 * 2) {
        this.customName = customName
    }

    override val type: InventoryType
        get() = InventoryType.FAKE_DOUBLE_CHEST

    override fun addViewer(player: Player, position: Vector, windowId: Byte) {
        viewer.add(player)
        if (openInventory.put(player, this) != null) {
            return
        }
        val blockPosition = onOpenChest(player)
        blockPositions[player] = blockPosition
        Server.instance.scheduler.scheduleDelayed(
            {
                val vector = blockPosition[0]
                val containerOpenPacket = ContainerOpenPacket()
                containerOpenPacket.id = windowId
                containerOpenPacket.uniqueEntityId = holderId
                containerOpenPacket.type = this.windowTypeId
                containerOpenPacket.blockPosition = vector.toVector3i()
                player.playerConnection.sendPacket(containerOpenPacket)
                this.sendContents(player)
            },
            3,
        )
    }

    override fun onOpenChest(player: Player): List<Vector> {
        val positionA = Vector(player.blockX, player.blockY + 2, player.blockZ)
        val positionB = positionA.add(1f, 0f, 0f)
        placeFakeChest(player, positionA)
        placeFakeChest(player, positionB)
        pair(player, positionA, positionB)
        pair(player, positionB, positionA)
        return listOf(positionA, positionB)
    }

    private fun pair(player: Player, positionA: Vector, positionB: Vector) {
        val blockEntityDataPacket = BlockEntityDataPacket()
        blockEntityDataPacket.blockPosition = positionA.toVector3i()
        blockEntityDataPacket.data = this.toChestNBT(positionA, positionB)
        player.playerConnection.sendPacket(blockEntityDataPacket)
    }

    private fun toChestNBT(positionA: Vector, positionB: Vector): NbtMap {
        return NbtMap.builder()
            .putString("id", "Chest")
            .putInt("x", positionA.blockX)
            .putInt("y", positionA.blockY)
            .putInt("z", positionA.blockZ)
            .putInt("pairx", positionB.blockX)
            .putInt("pairz", positionB.blockZ)
            .putString("CustomName", if (customName == null) "Chest" else customName)
            .build()
    }
}
